MovingBlocks/Terasology
# Refactor module installation code to decouple UI (SelectModulesScreen) from logic

### Contains

The title pretty much summarizes the content, the code to determine the required dependencies according to the [meta server](http://meta.terasology.org/modules/show) and download a module has been extracted from the `SelectModulesScreen` class. I did this for two reasons:
- As part of my GSoC project I'm implementing a feature in [FacadeServer](https://github.com/MovingBlocks/FacadeServer) to allow server administrators to manage multiple games on a server (start/stop the saved ones and create new games), and having the ability to install missing modules from the web interface is useful when you want to create a new game with a module which is currently not present on the server. As agreed with my mentor @msteiger redoing the dependency resolution and download code in FacadeServer would add useless duplication, so I extracted the existing code from the only user interface which uses it at the moment.
- `SelectModulesScreen` was becoming a huge class with multiple not-so-related inner classes (like MultiFileDownloader for example - it could later become useful for other purposes too).

### How to test

Using the module selection screen (which you can reach from Single Player > Create > Modules... ) should work as before, including installing new modules or updating existing ones. Some additional information is provided during the download (total bytes, downloaded bytes, number of modules, etc.); a small downside, however, is that when dependency resolution fails a generic not-very-detaild error message is shown. The reason is that the dependency resolution code has been refactored to use the already existing DependencyResolver class (with a ModuleRegistry based on the module list obtainde from the meta server), and providing additional information is an open issue (MovingBlocks/gestalt#15).

### Outstanding before merging

I have some doubts about the decision of which modules must be downloaded. I simply used the same strategy present in the old code, which is redownloading a dependency only if the online version is newer ([link to the old code](https://github.com/MovingBlocks/Terasology/blob/develop/engine/src/main/java/org/terasology/rendering/nui/layers/mainMenu/SelectModulesScreen.java#L485) - [link to the new code](https://github.com/MovingBlocks/Terasology/compare/develop...gianluca-nitti:refactor/module-downloader#diff-b0f5937519800e1a39bcbb713998c424R94)).
- Module dependencies have a range of compatible versions
- The Dependencyresolver returns a set of exact module versions
- This is then filtered to download only the versions newer than the installed ones

Is this correct? I did some manual tests and everything was fine, but I'm not 100% sure this is correct in any case of dependencies combinations, so I'd suggest someone with a deeper knowledge of the module dependency system reviews this before it's merged.
